crates/bevy_asset/src/lib.rs

```diff
@@ -1967,6 +1967,399 @@ mod tests {
+            let messages = collect_asset_events(world);
+            if messages.is_empty() {
+                return None;
+            }
+            assert_eq!(
+                messages,
+                [
+                    AssetEvent::LoadedWithDependencies { id: handle.id() },
+                    AssetEvent::Modified { id: handle.id() }
+                ]
+            );
+            Some(())
+        });
+    }
+
+    /// This is not a statement of intent but of behavior: One may load
+    /// their self path deferred of unknown type without error. It has the same
+    /// asset index as the original handle, but not the same type. And it can
+    /// reload.
+    ///
+    /// Caveat: Windows behaves strangely on this test. It loads a self-path
+    /// deferred with an unknown type, but once that path is loaded, it cannot
+    /// be retrieved from the `Assets<T>` resource.
+    #[test]
+    fn no_error_on_unknown_type_deferred_load_of_self_path() {
+        let (mut app, dir, source_events) = create_app_with_source_event_sender();
+        let asset_server = app.world().resource::<AssetServer>().clone();
+
+        dir.insert_asset_text(Path::new("abc.rsp"), "");
+
+        #[derive(Asset, TypePath)]
+        pub struct TestAssetUD(Handle<crate::LoadedUntypedAsset>);
+        struct ImmediateSelfLoader;
+
+        impl AssetLoader for ImmediateSelfLoader {
+            type Asset = TestAssetUD;
+            type Error = crate::loader::LoadDirectError;
+            type Settings = ();
+
+            async fn load(
+                &self,
+                _: &mut dyn Reader,
+                _: &Self::Settings,
+                load_context: &mut LoadContext<'_>,
+            ) -> Result<Self::Asset, Self::Error> {
+                let asset_path = load_context.path().clone();
+                let untyped_handle: Handle<crate::LoadedUntypedAsset> =
+                    load_context.loader().with_unknown_type().load(asset_path);
+
+                Ok(TestAssetUD(untyped_handle))
+            }
+
+            fn extensions(&self) -> &[&str] {
+                &["rsp"]
+            }
+        }
+
+        app.init_asset::<TestAssetUD>()
+            .register_asset_loader(ImmediateSelfLoader);
+
+        let handle: Handle<TestAssetUD> = asset_server.load("abc.rsp");
+
+        run_app_until(&mut app, |world| match asset_server.load_state(&handle) {
+            LoadState::Loading => None,
+            LoadState::Loaded => {
+                let asset_id = {
+                    let test_assets = world.resource::<Assets<TestAssetUD>>();
+                    let asset = test_assets.get(&handle).unwrap();
+                    asset.0.id()
+                };
+                let untyped = world.resource::<Assets<crate::LoadedUntypedAsset>>();
+                if let Some(untyped_handle) = untyped.get(asset_id) {
+                    assert_eq!(handle.id(), untyped_handle.handle.id());
+                    if cfg!(target_os = "windows") {
```

Comment:
I'm quite sure this is not a Windows problem - I think this is a flake-y test. We definitely shouldn't be cfg-ing out these asserts for Windows.

What's happening here is that the loader kicks off a **deferred** untyped load. That means its possible for the `TestAssetUD` to finish loading but for the untyped load to not even start. Hence you can have the TestAssetUD be loaded but the LoadedUntypedAsset not yet be loaded.

Instead of doing the assert here, use let-else to get the LoadedUntypedAsset and `return None;` if the let-else fails, and then do the assert afterwards.

Reply:
Thanks, @andriyDev! You're a wonder.